### DISCUSSION TITLE
Bug in examples/chat_app?

### DISCUSSION THREAD

**Discussion by kfalcami** [zauberzeug/nicegui#3841] [Q&A] [answered]
### Question

Hi,

I'm building an app pretty similar to chat_app, and started having problems where sometimes the input didn't show up in the chat, then usually the next input would cause both the missing and new input to show up.

I modified the original chat_app adding a time.sleep(x) in the send() function just before the messages.append call and it started having the same behavior.  On my laptop x=5 seconds makes it happen about half the time, x=8 almost every time.

This is surprising because I thought the chat_messages.refresh() call (which came after the sleep) would always refresh ui elements.  

Wrapping the sleep() call in "await run.io_bound(...)" and adding async to send() appears to solve the problem.

Questions:
1) why doesn't refresh() always work regardless of delay?  Some kind of race condition in there?
2) if there is a race, how would I guarantee that input is always displayed?  wrap the entire code for send() in run.io_bound maybe?

**Comment by python-and-novella** [comment]
plz provide a minimal  reproduce example.

**Reply by kfalcami** [reply]
It's the nicegui example chat_app, with a single line of code (sleep) added in the place described in my question.

**Comment by rodja** [comment, marked as answer]
> I modified the original chat_app adding a time.sleep(x) in the send() function

You should not block the main thread in an async application: https://github.com/zauberzeug/nicegui/wiki/FAQs#why-is-my-long-running-function-blocking-ui-updates
